src/native/clr/host/bridge-processing.cc

```diff
@@ -0,0 +1,307 @@
+	if (context->control_block->handle_type != JNIWeakGlobalRefType) [[unlikely]] {
+		log_error (LOG_DEFAULT, "Expected weak global reference type for handle, but got {} - handle: {:#x}", context->control_block->handle_type, reinterpret_cast<intptr_t> (context->control_block->handle));
+		return;
+	}
+	abort_unless (context->control_block->handle_type == JNIWeakGlobalRefType, "Expected weak global reference type for handle");
+
+	jobject weak = context->control_block->handle;
+	jobject handle = env->NewGlobalRef (weak);
+	
+	if (Logger::gref_log ()) [[unlikely]] {
+		OSBridge::_monodroid_gref_log (
+			std::format ("take_global_ref wref={:#x} -> handle={:#x}\n"sv,
+				reinterpret_cast<intptr_t> (weak),
+				reinterpret_cast<intptr_t> (handle)).data ());
+	}
+
+	if (handle != nullptr) {
+		context->control_block->handle = handle;
+		context->control_block->handle_type = JNIGlobalRefType;
+
+		OSBridge::_monodroid_gref_log_new (weak, OSBridge::get_object_ref_type (env, weak),
+				handle, OSBridge::get_object_ref_type (env, handle),
+				"finalizer", gettid (),
+				"   at [[clr-gc:take_global_ref]]", 0);
+
+		OSBridge::_monodroid_weak_gref_delete (weak, OSBridge::get_object_ref_type (env, weak),
+			"finalizer", gettid (), "   at [[clr-gc:take_global_ref]]", 0);
+		env->DeleteWeakGlobalRef (weak);
+	} else {
+		// The native memory of the control block will be freed in managed code as well as the weak global ref
+		context->control_block = nullptr;
+
+		if (Logger::gc_spew_enabled ()) [[unlikely]] {
+			OSBridge::_monodroid_gref_log (
+				std::format ("handle {:#x}/W; was collected by a Java GC"sv, reinterpret_cast<intptr_t> (weak)).data ());
+		}
+	}
+}
+
+void BridgeProcessing::take_weak_global_ref (HandleContext *context) noexcept
+{
+	abort_if_invalid_pointer_argument (context, "context");
+	abort_unless (context->control_block != nullptr, "Control block must not be null");
+	abort_unless (context->control_block->handle_type == JNIGlobalRefType, "Expected global reference type for handle");
+
+	jobject handle = context->control_block->handle;
+	if (Logger::gref_log ()) [[unlikely]] {
+		OSBridge::_monodroid_gref_log (std::format ("take_weak_global_ref handle={:#x}\n"sv, reinterpret_cast<intptr_t> (handle)).data ());
+	}
+
+	jobject weak = env->NewWeakGlobalRef (handle);
+	OSBridge::_monodroid_weak_gref_new (handle, OSBridge::get_object_ref_type (env, handle),
+			weak, OSBridge::get_object_ref_type (env, weak),
+			"finalizer", gettid (), "   at [[clr-gc:take_weak_global_ref]]", 0);
+
+	context->control_block->handle = weak;
+	context->control_block->handle_type = JNIWeakGlobalRefType;
+
+	OSBridge::_monodroid_gref_log_delete (handle, OSBridge::get_object_ref_type (env, handle),
+			"finalizer", gettid (), "   at [[clr-gc:take_weak_global_ref]]", 0);
+	env->DeleteGlobalRef (handle);
+}
+
+void BridgeProcessing::cleanup_after_java_collection () noexcept
+{
+#if DEBUG
+	int total = 0;
+	int alive = 0;
```

Comment:
Make the two variables `size_t`, that's the type of `scc.Count` and we can't really have negative counts here :)